File: codesLeftInWindows/CP-Problems-tilnoeneBackup/URI/2632.kt
import java.util.*
import java.lang.Math

fun min(a: Int, b: Int): Int{
    if(a < b) return a;
    return b;
}

fun dist(xa: Int, ya: Int, xb: Int, yb: Int): Int{
    return (xa-xb)*(xa-xb) + (ya-yb)*(ya-yb);
}

fun main(args: Array<String>) {
    val ler = Scanner(System.`in`)
    
    var n = ler.nextInt();
        
        for(i in 1..n){ 
            var ans = 0;
            var raio = 0;
            var dano = 0;
            
            var w = ler.nextInt();
            var h = ler.nextInt();
            var px0 = ler.nextInt();
            var py0 = ler.nextInt();
            
            var nome = ler.next();
            var nivel = ler.nextInt();
            var cx = ler.nextInt();
            var cy = ler.nextInt();
            
            var px1 = px0;
            var py1 = py0 + h;
                
            var px2 = px0 + w;
            var py2 = py0;
                
            var px3 = px0 + w;
            var py3 = py0 + h;
            
            if(nome == "fire"){
                dano = 200;
        
                if(nivel == 1){ 
                    raio = 20;
                } else if(nivel == 2){ 
                    raio = 30; 
                } else{ 
                    raio = 50; 
                }
        
            } else if(nome == "water"){
                dano = 300;
        
                if(nivel == 1){
                    raio = 10;
                } else if(nivel == 2){ 
                    raio = 25;
                } else{ 
                    raio = 40; 
                }
            } else if(nome == "earth"){
                dano = 400;
        
                if(nivel == 1){ 
                    raio = 25; 
                } else if(nivel == 2){ 
                    raio = 55;
                } else{ 
                    raio = 70; 
                }
            } else{
                dano = 100;
        
                if(nivel == 1){ 
                    raio = 18;
                } else if(nivel == 2){ 
                    raio = 38;
                } else{ 
                    raio = 60;
                }
            }
                
            if(dist(px0, py0, cx, cy) <= raio * raio){
                ans = dano;
            }
            if(dist(px1, py1, cx, cy) <= raio * raio){
                ans = dano;
            }
            if(dist(px2, py2, cx, cy) <= raio * raio){
                ans = dano;
            }
            if(dist(px3, py3, cx, cy) <= raio * raio){
                ans = dano;
            }
            
            if(ans == 0){
                if(cy < py3 && cy > py2){
                    if(min(dist(cx, cy, px2, cy), dist(cx, cy, px0, cy)) <= raio * raio){
                        ans = dano;
                    }
                }
                
                if(cx > px1 && cx < px3){
                    if(min(dist(cx, cy, cx, py1), dist(cx, cy, cx, py0)) <= raio * raio){
                        ans = dano;
                    }
                }
                
                if(cx >= px0 && cx <= px2 && cy >= py0 && cy <= py1){
                    ans = dano;
                }
            }
            
            
            println(ans);
        }
}

File: codesLeftInWindows/CP-Problems-tilnoeneBackup/URI/2632.lua
function split(inputstr, sep)
    if sep == nil then
            sep = "%s"
    end
    local t={}
    for str in string.gmatch(inputstr, "([^"..sep.."]+)") do
            table.insert(t, str)
    end
    return t
end

function min(a, b)
    if a < b then return a end
    return b
end

function dist(xa, ya, xb, yb)
    return (xa-xb)*(xa-xb) + (ya-yb)*(ya-yb)
end

    n = tonumber(io.read())
    
    for i=1,n do
        ans = 0
        
        k = io.read()
        l = split(k)
        
        w = tonumber(l[1])
        h = tonumber(l[2])
        px0 = tonumber(l[3])
        py0 = tonumber(l[4])
        
        px1 = px0
        py1 = py0 + h
        
        px2 = px0 + w
        py2 = py0
        
        px3 = px0 + w
        py3 = py0 + h
        
        k = io.read()
        l = split(k)
        
        nome = l[1]
        nivel = tonumber(l[2])
        cx = tonumber(l[3])
        cy = tonumber(l[4])
        
        if(nome == "fire") then
            dano = 200

            if(nivel == 1) then raio = 20
            elseif(nivel == 2) then raio = 30
            else raio = 50 end

        elseif(nome == "water") then
            dano = 300

            if(nivel == 1) then raio = 10
            elseif(nivel == 2) then raio = 25
            else raio = 40 end

        elseif(nome == "earth") then
            dano = 400

            if(nivel == 1) then raio = 25
            elseif(nivel == 2) then raio = 55
            else raio = 70 end

        else
            dano = 100

            if(nivel == 1) then raio = 18
            elseif(nivel == 2) then raio = 38
            else raio = 60 end
        end
            
        if(dist(px0, py0, cx, cy) <= raio * raio) then
            ans = dano end
        if(dist(px1, py1, cx, cy) <= raio * raio) then
            ans = dano end
        if(dist(px2, py2, cx, cy) <= raio * raio) then
            ans = dano end
        if(dist(px3, py3, cx, cy) <= raio * raio) then
            ans = dano end
        
        if(ans == 0) then
            if(cy < py3 and cy > py2) then
                if(min(dist(cx, cy, px2, cy), dist(cx, cy, px0, cy)) <= raio * raio) then
                    ans = dano
                end
            end

            if(cx > px1 and cx < px3) then
                if(min(dist(cx, cy, cx, py1), dist(cx, cy, cx, py0)) <= raio * raio) then
                    ans = dano
                end
            end

            if(cx >= px0 and cx <= px2 and cy >= py0 and cy <= py1) then
                ans = dano
            end
    end    
    
        print(ans)
    end